loki/frontend/fparser.py

```diff
@@ -1825,24 +1863,11 @@ def visit_Subroutine_Subprogram(self, o, **kwargs):
             spec._update(body=spec.body + body.body[:idx])
             body._update(body=body.body[idx:])
 
-        # Another big hack: fparser allocates all comments before and after the
-        # spec to the spec. We remove them from the beginning to get the docstring.
-        comment_map = {}
-        docs = []
-        for node in spec.body:
-            if not isinstance(node, (ir.Comment, ir.CommentBlock)):
-                break
-            docs.append(node)
-            comment_map[node] = None
+        # Extract the leading comments of the specification as "docstring" section
+        docs = _get_comments_from_section(spec) if spec else ()
 
-        # Now we move comments from the end to the beginning of the body as those
-        # can potentially be pragmas.
-        for node in reversed(spec.body):
-            if not isinstance(node, (ir.Pragma, ir.Comment, ir.CommentBlock)):
-                break
-            body.prepend(node)
-            comment_map[node] = None
-        spec = Transformer(comment_map, invalidate_source=False).visit(spec)
+        # Move trailing comments from spec to the body as those can be pragmas.
```

Comment:
[no action only and observation] The comment's not 100% accurate anymore. We're moving comments and directives, but the latter are now identified as different node types (which wasn't the case at the time this was first built, iirc). The conceptual step is still required, though, and we can't bail out early either, because you might have interleaved comments and preprocesser directives and you wouldn't want to miss any of the latter